src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs

```diff
@@ -2209,8 +2209,7 @@ internal void DeriveParameters()
                         Debug.Assert(_activeConnection.IsKatmaiOrNewer, "Invalid datatype token received from pre-katmai server");
 
                         //read the type name
-                        p.TypeName = r[colNames[(int)ProcParamsColIndex.TypeCatalogName]] + "." +
-                            r[colNames[(int)ProcParamsColIndex.TypeSchemaName]] + "." +
+                        p.TypeName = r[colNames[(int)ProcParamsColIndex.TypeSchemaName]] + "." +
```

Comment:
I suspect this fix could break some use cases. Though I have not looked at the bug being mentioned and neither have I explored this API, in detail, but the change implies that we are removing CatalogName from the TypeName. 
The code being modified doesn't seem like something only used for TVPs. 

Have you done any analysis on what could break if this is changed to your proposed change? Or any reasoning about how nothing would break, if this code is changed? 

Is there a possibility that this fix could only be applied to the specific breaking scenario, so as to reduce the risk?

Reply:
The analysis I have done is not super in depth, however there are two points that led me to believe that this specific code is specific to TVPs:

- `DeriveParameters()` is wholly restricted to commands calling stored procedures. In the beginning of the function, if the CommandType is anything other than StoredProcedure the function calls an exception
https://github.com/dotnet/corefx/blob/f9b470ff1bacf44dc58d967931e962a30ba32d41/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs#L2011-L2022
- The conditional branch that the changed code is in specifically refers to `SqlDbType.Structured` which is defined as 30 in SqlDbType.cs - this correlates to the TVP when I run `sp_procedure_params_100_managed` on my test sproc.

I am not familiar enough with the inner workings of SQL Server to know if the MANAGED_DATA_TYPE of 30 is used for anything other than TVPs, but overall it appears this code is isolated to the type name of the TVP being passed in to a stored procedure.

I am happy to do further analysis if needed, please let me know your thoughts.

Reply:
I think your explanation is reasonable and I am leaning towards taking this fix. 
I do think that some more due diligence is needed from our side (SqlClient committers) 
I am going to do a bit more investigation from my end to see if I can find some tests which test something special in this code path (from the .Net framework repository). I will try to see if someone from Sql Server can explain the MANAGED_DATA_TYPE codes and its history. 

**Backward compatiblity**
Even if we take this fix, I have a concern. I looked at your repo where you have tests which repro the bug and also a repro which incorporates a workaround, and also found a link to the StackOverFlow post from 2012. Imagine a scenario where an app, with the workaround, migrates to this "fixed" version of SqlClient. We can assume that the migration happened from the previous versions of SqlClient / or from .Net Framework. 
With this fix, a bug will now be introduced in those applications as the workaround starts breaking. We could add a quirk for this fix, so that either it is an opt-in fix, or an opt-out fix. My guess would be that since this bug has probably been around since TVP support was added to SqlClient, all customers depending on this API would have a workaround in place and this probably needs to be an opt-in fix. 

I would need more inputs here. @karelz and @danmosemsft can you help with a decision regarding backward compat in this case? Am I thinking too much, or is this backward compatiblity issue a real concern?

Reply:
Start with a sproc that takes a tvp param and goes through the broken code producing a 3-part name for the param. We're now in a broken position. Now the user can intervene in their own code by altering the parameter name so we have a 2 part name. Or they can not and we have a 3-part name.

Now the user causes the command object to execute and we have to create the rpc structures and populate the parameters and their values into tds stream. At this point we don't hand back control flow to the user again until we're done sending the command so we're in charge of state. While writing the params can't we do part of the parse step again (i have a span based version i was working on somewhere) and if we find a 3-part name simply fix it up?

It would avoid the user having to intervene but if they had code which already did then we wouldn't alter their work. Essentially we don't do this fix, we do it later in the execution pipeline at a stage when there's no way the user can (legally) alter the parameter object. It would mean existing code kept working but new code would silently fix itself allowing new code to work.

Reply:
> While writing the params can't we do part of the parse step again

What will you parse here?

Reply:
>nd if we find a 3-part name simply fix it up?
> What will you parse here?

Can you provide and example of how the input will be parsed and what the fix up will look like?

Reply:
@Wraith2 I like where you are going - basically we can implement the original workaround within something like ExecuteNonQueryInternal - if the command type is StoredProcedure, check any Structured parameters for a 3-part name. If they have a 3-part name run the workaround, otherwise continue on as normal. Addresses the backwards compatibility issue and fixed the bug.

Reply:
It's late and I've not done the full rundown (you know it's complicated in there) but I think [this](https://github.com/dotnet/corefx/blob/5d2745a3af6e2b20a1dcb36b9ff5f72a59d08697/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs#L3310) looks like a reasonable place to put it. in that if block add a call to the multipartidentifier parse function with the type name and fixup as needed.

Reply:
More than where we are going to put the fix, I want to understand what the fix is? 

Is it going to rely on tokenizing the name by "." and figure out if it is a 3 part name? If Structured type, then remove the first token and pass it down to the server?

Reply:
In that case, there is no guarantee the part before the first "." is the database name or not. This is because the TVP name could be "Company.MyTVP"
In case the input is dbo.Company.MyTVP then the above solution will end up stripping off dbo

If the TVP name is Company.MyTVP.Segment1 then the Company portion of the TVP name will be stripped. 

A compat change issue with the general approach above, is that, there were a set of inputs, which could result in an error and now it is not resulting in an error. 

Let's say that the inputs were deliberately meant to be DBName.dbo.TVPName (not retrieved from DeriveParameters) and a SQLException was thrown. Now adding a fix-up logic, changes the behavior.

Reply:
As far as i know TVP name of `DBName.dbo.TVPName` isn't legal,  If this isn't the case then the error that is currently thrown (incorrectly) is itself incorrect and needs to be reviewed. So we need a definitive answer on whther TVP names can be 3 or 4 part or if they must be exactly 2 parts.

The TVP name can't be `Company.MyTVP.Segment1` it would have to be `Company.[MyTVP.Segment1]`. This is why i said you need to call ParseMultipartIdentifier. It contains a parser which handles all the naming rules and it's the only safe way to determine how many parts there are. It would pick up `Company.MyTVP.Segment1` as 3 part name and throw.

Reply:
I understand now and yes you're right we'd need a flag on the parameter object to determine if the typename had come from DeriveParameters and not been altered by the user so we could determine if we should fixup our own mistake or throw because the user made it.